File: deob/src/main/java/dev/openrs2/deob/transform/FieldOrderTransformer.java
package dev.openrs2.deob.transform;

import java.util.HashMap;

import dev.openrs2.asm.MemberDesc;
import dev.openrs2.asm.classpath.ClassPath;
import dev.openrs2.asm.classpath.Library;
import dev.openrs2.asm.transform.Transformer;
import org.objectweb.asm.Opcodes;
import org.objectweb.asm.tree.ClassNode;
import org.objectweb.asm.tree.FieldInsnNode;

public final class FieldOrderTransformer extends Transformer {
	private static final String CONSTRUCTOR = "<init>";
	private static final String STATIC_CONSTRUCTOR = "<clinit>";

	private static void sortFields(ClassNode clazz, String ctorName, int opcode) {
		clazz.methods.stream()
			.filter(m -> m.name.equals(ctorName))
			.findFirst()
			.ifPresent(ctor -> {
				var fields = new HashMap<MemberDesc, Integer>();
				var index = 0;

				for (var it = ctor.instructions.iterator(); it.hasNext(); ) {
					var insn = it.next();
					if (insn.getOpcode() != opcode) {
						continue;
					}

					var putfield = (FieldInsnNode) insn;
					if (!putfield.owner.equals(clazz.name)) {
						continue;
					}

					var desc = new MemberDesc(putfield);
					if (!fields.containsKey(desc)) {
						fields.put(desc, index++);
					}
				}

				clazz.fields.sort((a, b) -> {
					var indexA = fields.getOrDefault(new MemberDesc(a), -1);
					var indexB = fields.getOrDefault(new MemberDesc(b), -1);
					return Integer.compare(indexA, indexB);
				});
			});
	}

	@Override
	public boolean transformClass(ClassPath classPath, Library library, ClassNode clazz) {
		sortFields(clazz, CONSTRUCTOR, Opcodes.PUTFIELD);
		sortFields(clazz, STATIC_CONSTRUCTOR, Opcodes.PUTSTATIC);
		return false;
	}
}

File: deob/src/main/java/dev/openrs2/deob/transform/FieldOrderTransformer.kt
package dev.openrs2.deob.transform

import dev.openrs2.asm.MemberDesc
import dev.openrs2.asm.classpath.ClassPath
import dev.openrs2.asm.classpath.Library
import dev.openrs2.asm.transform.Transformer
import org.objectweb.asm.Opcodes
import org.objectweb.asm.tree.ClassNode
import org.objectweb.asm.tree.FieldInsnNode
import org.objectweb.asm.tree.FieldNode
import javax.inject.Singleton

@Singleton
class FieldOrderTransformer : Transformer() {
    override fun transformClass(classPath: ClassPath, library: Library, clazz: ClassNode): Boolean {
        sortFields(clazz, CONSTRUCTOR, Opcodes.PUTFIELD)
        sortFields(clazz, STATIC_CONSTRUCTOR, Opcodes.PUTSTATIC)
        return false
    }

    private companion object {
        private const val CONSTRUCTOR = "<init>"
        private const val STATIC_CONSTRUCTOR = "<clinit>"
        private val STATIC_COMPARATOR = Comparator<FieldNode> { a, b ->
            val aStatic = a.access and Opcodes.ACC_STATIC != 0
            val bStatic = b.access and Opcodes.ACC_STATIC != 0
            when {
                aStatic && !bStatic -> -1
                !aStatic && bStatic -> 1
                else -> 0
            }
        }

        private fun sortFields(clazz: ClassNode, ctorName: String, opcode: Int) {
            val ctor = clazz.methods.find { it.name == ctorName } ?: return

            val fields = mutableMapOf<MemberDesc, Int>()
            var index = 0
            for (insn in ctor.instructions) {
                if (insn.opcode != opcode) {
                    continue
                }

                val putfield = insn as FieldInsnNode
                if (putfield.owner != clazz.name) {
                    continue
                }

                val desc = MemberDesc(putfield)
                if (!fields.containsKey(desc)) {
                    fields[desc] = index++
                }
            }

            clazz.fields.sortWith(STATIC_COMPARATOR.thenBy { fields.getOrDefault(MemberDesc(it), -1) })
        }
    }
}
